File: backend/application/document_parser.py
import io
import logging

import pdfplumber
from docx import Document

logger = logging.getLogger(__name__)

MAX_TEXT_LENGTH = 30_000


def extract_text_from_pdf(file_bytes: bytes) -> str:
    text_parts = []
    with pdfplumber.open(io.BytesIO(file_bytes)) as pdf:
        for page in pdf.pages:
            page_text = page.extract_text()
            if page_text:
                text_parts.append(page_text)
    return "\n".join(text_parts).strip()


def extract_text_from_docx(file_bytes: bytes) -> str:
    doc = Document(io.BytesIO(file_bytes))
    text_parts = []
    for para in doc.paragraphs:
        if para.text.strip():
            text_parts.append(para.text.strip())
    return "\n".join(text_parts).strip()


def extract_text(filename: str, file_bytes: bytes) -> str:
    lower = filename.lower()
    if lower.endswith(".pdf"):
        text = extract_text_from_pdf(file_bytes)
    elif lower.endswith(".docx"):
        text = extract_text_from_docx(file_bytes)
    else:
        raise ValueError("Unsupported file type. Please upload a PDF or DOCX file.")

    if not text.strip():
        raise ValueError("Could not extract text from the file. It may be image-based or empty.")

    if len(text) > MAX_TEXT_LENGTH:
        text = text[:MAX_TEXT_LENGTH]

    return text

File: backend/src/services/document-parser.js
import { PDFParse } from "pdf-parse";
import mammoth from "mammoth";

const MAX_TEXT_LENGTH = 30_000;

export async function extractText(filename, fileBytes) {
  const lower = filename.toLowerCase();
  let text;

  if (lower.endsWith(".pdf")) {
    const parser = new PDFParse(new Uint8Array(fileBytes));
    await parser.load();
    text = (await parser.getText()).trim();
  } else if (lower.endsWith(".docx")) {
    const result = await mammoth.extractRawText({ buffer: fileBytes });
    text = result.value.trim();
  } else {
    throw new Error("Unsupported file type. Please upload a PDF or DOCX file.");
  }

  if (!text) {
    throw new Error("Could not extract text from the file. It may be image-based or empty.");
  }

  if (text.length > MAX_TEXT_LENGTH) {
    text = text.slice(0, MAX_TEXT_LENGTH);
  }

  return text;
}
